### DISCUSSION TITLE
How to make a "resettable" `RefCountAsyncDisposable`

### DISCUSSION THREAD

**Discussion by gmkado** [dotnet/reactive#2149] [Q&A] [answered]
Pardon the long-winded writeup.  This isn't regarding anything related to Rx but just something I came across when trying to use `System.Reactive.Disposables.RefCountAsyncDisposable` so just thought I'd post my findings.

I have a scenario where I want to create a "scope" that can be entered in a multi-threaded environment.  The first time the scope is entered, it performs some asynchronous entry action, then when all the scopes are left it performs some asynchronous exit action.  After that it resets so the next entry calls the entry action, etc.

`RefCountAsyncDisposable` seemed to fit the bill, not sure if this is meant to be an internal utility class but its exposed as public so I'm trying it out.  Here is my naive first approach:

```cs
private RefCountAsyncDisposable? refCounted;
private readonly AsyncGate gate = new();

async Task<IAsyncDisposable> GetScopeBad()
{
    using var _ = await gate.LockAsync();
        
    if(refCounted == null) 
    {
        /* do entry action */
        refCounted = new RefCountAsyncDisposable(AsyncDisposable.Create(async () => {
            /* do exit action */
            refCounted = null;
        }));
        var scope = await refCounted.GetDisposableAsync();
        await refCounted.DisposeAsync();
        return scope;        
    }            
    // Race condition here if previous scopes are now exited
    return await refCounted.GetDisposableAsync();
}
```
This works for the most part, the first time around `refCounted` is created, the scope is retrieved with `GetDisposableAsync`, and `refCounted` is immediately disposed so that as soon as the ref counts go to zero it will be disposed.  In the dispose callback, `refCounted` is set back to null so the cycle can repeat.  I use an `AsyncGate` to avoid race conditions while setting `refCounted`.

I can think of one race condition where we've entered the scope previously so `refCounted` is not null and we skip resetting it.  But before we get to the last line of the method, that other scope is exited and `refCounted` is set to null and we get a NPE.  Locking the dispose callback avoids the NPE but won't prevent the scope from being disposed early.

Here is my second approach:

```cs
async Task<IAsyncDisposable> GetScopeGood()
{
    using var _ = await gate.LockAsync();
    
    var scope = refCounted != null ? await refCounted.GetDisposableAsync() : AsyncDisposable.Nop;    
    if(scope == AsyncDisposable.Nop) 
    {
        /* do entry action */
        refCounted = new RefCountAsyncDisposable(AsyncDisposable.Create(async () => {
            /* do exit action */
        }));

        scope = await refCounted.GetDisposableAsync();
        await refCounted.DisposeAsync();
        return scope;        
    }            
    // Now if previous scopes are exited, the scope obtained above will keep refCounted from disposing
    return scope;
}
```

I believe this covers the aforementioned race condition **but** relies on `GetDisposableAsync` returning `Nop` if already disposed, and for all `Nop` disposables to be equal, which seems like an implementation detail that could change in the future.

Am I overthinking this?  Is there something better suited for this kind of thing?  

Here's my full interactive c# playground demoing the race condition:

```cs
#r "nuget: System.Reactive.Async, 6.0.0-alpha.18"
#nullable enable

using System.Reactive.Disposables;
using System.Threading.Tasks;
using System.Threading;
using System;

RefCountAsyncDisposable? refCounted;
AsyncGate gate = new();

delegate Task<IAsyncDisposable> GetScopeDelegate(IAsyncDisposable? previous = null);

async Task SimulateRaceCondition(IAsyncDisposable previous)
{
    Console.WriteLine("Simulating race condition");
    using(ExecutionContext.SuppressFlow())
    {
        // need to suppress flow for async gate to simulate properly, but Console.WriteLine will no longer work (???)
        Task.Run(previous.DisposeAsync);  
    }
    // Some time to make sure its been called if its going to be called
    await Task.Delay(100);
    Console.WriteLine("Done simulating race condition");
}

async Task<IAsyncDisposable> GetScopeBad(IAsyncDisposable? previous = null)
{
    using var _ = await gate.LockAsync();
        
    if(refCounted == null) 
    {
        Console.WriteLine("refCount reset");
        refCounted = new RefCountAsyncDisposable(AsyncDisposable.Create(async () => {
            using var _ = await gate.LockAsync();            
            refCounted = null;
            Console.WriteLine("dispose complete");            
        }));
        var scope = await refCounted.GetDisposableAsync();
        await refCounted.DisposeAsync();
        return scope;        
    }            

    if(previous != null) await SimulateRaceCondition(previous);
    return await refCounted.GetDisposableAsync();
}

async Task<IAsyncDisposable> GetScopeGood(IAsyncDisposable? previous = null)
{

    using var _ = await gate.LockAsync();
    
    var scope = refCounted != null ? await refCounted.GetDisposableAsync() : AsyncDisposable.Nop;    
    if(scope == AsyncDisposable.Nop) 
    {
        Console.WriteLine("refCount reset");
        refCounted = new RefCountAsyncDisposable(AsyncDisposable.Create(async () => {                        
            using var _ = await gate.LockAsync();
            Console.WriteLine("dispose complete");
        }));
        scope = await refCounted.GetDisposableAsync();
        await refCounted.DisposeAsync();
        return scope;        
    }            

    if(previous != null) await SimulateRaceCondition(previous);
    return scope;
}

GetScopeDelegate GetScope = GetScopeBad;
// GetScopeDelegate GetScope = GetScopeGood;

async Task PrintDisposeState()
{
    var scope = refCounted != null ? await refCounted.GetDisposableAsync() : AsyncDisposable.Nop;    
    Console.WriteLine($"refCounted is {(scope == AsyncDisposable.Nop ? "disposed": "not disposed")}");
    await scope.DisposeAsync();
}

async Task Run()
{
    Console.WriteLine("Acquiring 1...");
    var r1 = await GetScope();

    Console.WriteLine("Acquiring 2 and disposing 1 (race condition!)...");
    var r2 = await GetScope(r1);
    
    await PrintDisposeState();

    Console.WriteLine("Disposing 2");
    await r2.DisposeAsync();
}

await Run();
Console.WriteLine("At this point we should be disposed...");
await PrintDisposeState();

Console.WriteLine("Running again...");
await Run();
```

BAD (without lock):

```
Acquiring 1...
refCount reset
Acquiring 2 and disposing 1 (race condition!)...
Simulating race condition
Done simulating race condition
Error: System.NullReferenceException: Object reference not set to an instance of an object.
```

BAD (with lock):

```
Acquiring 1...
refCount reset
Acquiring 2 and disposing 1 (race condition!)...
Simulating race condition
Done simulating race condition
refCounted is disposed <----- OOPS
Disposing 2
At this point we should be disposed...
refCounted is disposed
Running again...
Acquiring 1...
refCount reset
Acquiring 2 and disposing 1 (race condition!)...
Simulating race condition
Done simulating race condition
refCounted is disposed <----- OOPS
Disposing 2
```

GOOD:

```
Acquiring 1...
refCount reset
Acquiring 2 and disposing 1 (race condition!)...
Simulating race condition
Done simulating race condition
refCounted is not disposed <----- FIXED
Disposing 2
dispose complete
At this point we should be disposed...
refCounted is disposed
Running again...
Acquiring 1...
refCount reset
Acquiring 2 and disposing 1 (race condition!)...
Simulating race condition
Done simulating race condition
refCounted is not disposed <----- FIXED
Disposing 2
dispose complete
```

**Comment by idg10** [comment, marked as answer]
Although it's possible to create an `IDiposable` that can be undisposed, so to speak, it's not the norm, and it presents some dangers.

`IDisposable` is an instance of a more general pattern of a "one-shot" kind of usage model. `Task` is another. From the consumer's perspective, these types will do their job and then you can't do anything more with them. If you've got an `IDisposable` object then the normal rule are that you can't use it again after you call `Dispose`. If you've got a `Task`, it completes once, then remains in a completed state and you can't ask it to run again.

The reason I bring up the general pattern, and `Task` in particular is that `ValueTask` sheds some light on what happens if you do actually try to reuse these things. The basic purpose of `ValueTask` is to reduce allocations on the GC heap by making it possible to reuse underlying `Task`s in some situations. This is highly effective, but it also makes the programming model of `ValueTask` much more restrictive: you can only `await` it once. And the reason for that is that a `ValueTask` may well be a wrapper around a `Task`-like thing that will actually be reset and reused.

The danger is that if you misuse a `ValueTask` by `await`ing it twice, the second `await` might now complete when some completely unrelated piece of work finishes (because the underlying `Task`-like thing got recycled and attached to some completely different piece of work). The insidious thing about this is that you don't get an obvious indication that anything is wrong. `ValueTask` doesn't throw an exception if you `await` it twice. The whole point of this type is that it's a lightweight proxy for a resettable object that may be reused in other contexts, and for value task to be able to detect that you've made this mistake would effectively require an allocated object on the heap per `ValueTask`, defeating its purpose.

To generalize that: if we take things that are normally one-shot and make them resettable, we introduce the potential for a bug that can't be detected at runtime. If an object is used after being reset it has no way of knowing whether that's a correct use from its new owner, or it's the original owner incorrectly trying to use it after it is no longer the owner.

So for this reason I don't want to make any of our `IDiposable` implementations resettable. Today they are all one-shot implementations, meaning they do not create opportunities for this form of bug.

Also, the primary purpose of the Rx Disposables is to do what Rx needs. Since they are public, we will fully support their existing API into the future. And that commitment cuts both ways: since we intend to support all public features of these types, that means adding new capabilities increases ongoing maintenance overheads. So for that reason, we don't really want to be expanding the feature set of these types outside of what Rx itself need from them.

That's not an absolutely inflexible position. I would always consider a request for new functionality, and there may well be cases where we would actually add something new. However in this particular case, the change from "one shot" to "resettable" is a big change, and as I've described above, one I consider to enable somewhat risky patterns.

For the problem you're trying to solve, I my first thought would have been an `int` and using `Interlocked.Increment`/`Decrement` as callers enter and leave, and this would give you a way of detecting every time that it drops to zero in a way that is thread safe. The only tricky part would be whether you wanted to make any particular guarantees about the execution of the end-of-scope callback. A naive implementation of the kind I describe could result in that callback running multiple times concurrently if the scope is being entered and left frequently and the callback is slow. But if you want to prevent that I think you'd need to think quite carefully about how you do actually want to handle that. It's not obvious to me what a good general concurrent-call-preventing design would be.

**Comment by gmkado** [comment]
Once again thanks @idg10 for the detailed response!  I posted this as a discussion and not an issue because I don't want to make any changes to existing Rx Disposables API, but I wanted to discuss how I'm using the public API for this specific use case.

Using a thread-safe counter is much simpler than what I had using `RefCountAsyncDisposable`, thanks for the suggestion.  I locked the counter and the callback together to avoid the issue you mentioned in the last paragraph, for my application it makes sense for callers to wait until `_onEnter` and `_onExit` have been fully completed.  

Here's what I ended up with:

```cs
    public async Task<IAsyncDisposable> GetScopeAsync(CancellationToken cancellationToken)
    {
        using (await _gate.LockAsync())
        {
            _counter++;
            if(_counter == 1)
            {
                await _onEnter(cancellationToken);
            }
        }

        return AsyncDisposable.Create(async ()=>
        {
            using var _ = await _gate.LockAsync();
            _counter--;
            if(_counter == 0)
            {
                await _onExit();
            }
        });
    }
```

**Reply by idg10** [reply]
One thing I would be slightly wary of in this design is that you're holding onto an internal lock while calling out into code you might not control. If that `_onEnter` call takes a long time, nothing else can enter the scope until it's done. Similarly, the scope will be unavailable for anyone to enter while `_onExit` is in progress.

That might be OK. (Rx's `Synchronized` essentially has the same characteristic. In fact that's basically the point of `Synchronized`.) But it might also cause problems. Holding onto a lock while calling out into application code can make it easy to introduce deadlocks. So I can imagine scenarios in which you might want to decouple these scope enter/exit notifications from the actual acquisition and release of scopes. But I can also imagine scenarios where this simple approach is entirely appropriate.

That's what I meant when I said "it's not obvious to me what a good general concurrent-call-preventing design would be." I can imagine multiple scenarios, and no single design is the best fit for them all. I don't know what you're doing in those `_onEnter` and `_onExit` calls, so I don't know the constraints that apply in your particular scenario.
